src/servers/src/prom_row_builder.rs

```diff
@@ -0,0 +1,249 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::string::ToString;
+
+use api::prom_store::remote::Sample;
+use api::v1::value::ValueData;
+use api::v1::{
+    ColumnDataType, ColumnSchema, Row, RowInsertRequest, RowInsertRequests, Rows, SemanticType,
+    Value,
+};
+use common_query::prelude::{GREPTIME_TIMESTAMP, GREPTIME_VALUE};
+
+use crate::proto::PromLabel;
+use crate::repeated_field::Clear;
+
+#[derive(Default)]
+pub(crate) struct TablesBuilder {
+    tables: HashMap<String, TableBuilder>,
+}
+
+impl Clear for TablesBuilder {
+    fn clear(&mut self) {
+        self.tables.clear();
+    }
+}
+
+impl TablesBuilder {
+    pub(crate) fn get_or_create_table_builder(&mut self, table_name: String) -> &mut TableBuilder {
+        self.tables.entry(table_name).or_default()
+    }
+
+    pub(crate) fn as_insert_requests(&mut self) -> (RowInsertRequests, usize) {
+        let mut total_rows = 0;
+        let inserts = self
+            .tables
+            .drain()
+            .map(|(name, mut table)| {
+                total_rows += table.num_rows();
+                table.as_row_insert_request(name)
+            })
+            .collect();
+        (RowInsertRequests { inserts }, total_rows)
+    }
+}
+
+pub(crate) struct TableBuilder {
+    schema: Vec<ColumnSchema>,
+    rows: Vec<Row>,
+    col_indexes: HashMap<String, usize>,
+}
+
+impl Default for TableBuilder {
+    fn default() -> Self {
+        let col_indexes = [
+            (GREPTIME_TIMESTAMP.to_string(), 0),
+            (GREPTIME_VALUE.to_string(), 1),
+        ]
+        .into_iter()
+        .collect();
+        Self {
+            schema: vec![
+                ColumnSchema {
+                    column_name: GREPTIME_TIMESTAMP.to_string(),
+                    datatype: ColumnDataType::TimestampMillisecond as i32,
+                    semantic_type: SemanticType::Timestamp as i32,
+                    datatype_extension: None,
+                },
+                ColumnSchema {
+                    column_name: GREPTIME_VALUE.to_string(),
+                    datatype: ColumnDataType::Float64 as i32,
+                    semantic_type: SemanticType::Field as i32,
+                    datatype_extension: None,
+                },
+            ],
+            rows: vec![],
+            col_indexes,
+        }
+    }
+}
+
+impl TableBuilder {
+    fn num_rows(&self) -> usize {
+        self.rows.len()
+    }
+    pub(crate) fn add_labels_and_samples(&mut self, labels: &[PromLabel], samples: &[Sample]) {
+        let mut row = vec![Value { value_data: None }; self.col_indexes.len()];
+
+        for PromLabel { name, value } in labels {
+            let tag_name = unsafe { String::from_utf8_unchecked(name.to_vec()) };
+            let tag_value = unsafe { String::from_utf8_unchecked(value.to_vec()) };
```

Comment:
nitpick: Some docs for unsafe.

Reply:
fixed https://github.com/GreptimeTeam/greptimedb/pull/3425/commits/571926fd0e2cda3f14633107fa722bf6e73c55c8#diff-7c0d742ac3a9cfc88c110b929c9f0a43f63cf981264415ff10c25e0710749ed7R124